xapian-core/include/xapian/cluster.h

```diff
@@ -0,0 +1,325 @@
+    //Vector storing the documents for this DocumentSet
+    std::vector<Document> docs;
+  public:
+    // Returns the size of the DocumentSet
+    int size() {
+	return docs.size();
+    }
+
+    // Returns the document at index i
+    Xapian::Document operator[](Xapian::doccount i);
+
+    /* Returns an iterator to the start of the DocumentSet to iterate
+     * through all the documents
+     */
+    DocumentSetIterator begin();
+
+    // Returns an iterator to the end of the DocumentSet
+    DocumentSetIterator end();
+
+    // Adds a document to the DocumentSet
+    void add_document(Document &doc);
+};
+
+// A class used to iterate through the DocumentSet
+class XAPIAN_VISIBILITY_DEFAULT DocumentSetIterator {
+    friend class DocumentSet;
+    DocumentSet docs;
+    int index;
+  public:
+    DocumentSetIterator(DocumentSet &docs_, int index_)
+        : docs(docs_), index(index_) {}
+
+    // Returns the document referred to by the current iterator position
+    Xapian::Document get_document();
+
+    // Prefix increment operator for this iterator
+    DocumentSetIterator & operator++() {
+	index++;
+	return *this;
+    }
+
+    // Postfix increment operator for this iterator
+    DocumentSetIterator operator++(int) {
+	DocumentSetIterator d = *this;
+	index++;
+	return d;
+    }
+
+    // Prefix decrement operator for this iterator
+    DocumentSetIterator & operator--() {
+	index--;
+	return *this;
+    }
+
+    // Postfix decrement operator for this iterator
+    DocumentSetIterator operator--(int) {
+	DocumentSetIterator d = *this;
+	index--;
+	return d;
+    }
+};
+
+// Base class used for construction of the TermLists
+class XAPIAN_VISIBILITY_DEFAULT FreqSource {
+  public:
+    /** This contains a map of the terms and its corresponding term frequencies.
+     *  The term frequency of a term stands for the number of documents it indexes
+     */
+    std::map<std::string, Xapian::doccount> termfreq;
+
+    // Destructor
+    virtual ~FreqSource();
+
+    // Gets the term frequency of a particular term 'tname'
+    virtual Xapian::doccount get_termfreq(const std::string &tname) = 0;
+
+    // Gets the number of documents
+    virtual Xapian::doccount get_doccount() = 0;
+};
+
+// A class for dummy frequency source for construction of termlists
+class XAPIAN_VISIBILITY_DEFAULT DummyFreqSource : public FreqSource {
+  public:
+    Xapian::doccount get_termfreq(const std::string &tname);
+
+    Xapian::doccount get_doccount();
+};
+
+// A class for construction of termlists
+class XAPIAN_VISIBILITY_DEFAULT TermListGroup : public FreqSource {
+  public:
+    /** Adds document terms and their statistics to the FreqSource for
+     *  construction of termlists
+     */
+    doccount docs_num;
+
+    void add_document(const Document &doc);
+
+    // Add a number of documents from the DocumentSource
+    void add_documents(DocumentSource &docs);
+
+    Xapian::doccount get_termfreq(const std::string &tname);        
```

Comment:
This has trailing whitespace. (I found this by looking at the output of `git diff`, where it's highlighted. That's also used by `git add -p`, which is one of the reasons I recommend adding things into the git index using it.)